### PR TITLE
feat: Add ListMetadata + Read + Write functionality to Attio connector

### PR DESCRIPTION
**Pull request by sanjay-a01**
Response:
image

### CODE REVIEW

File: providers/attio/metadata.go
Comment (Cobalt0s) [inline providers/attio/metadata.go:111]: Since we are only collecting "keys", `ajson` library has method that will avoid additional marshall, unmarshall. It would look like this:
```
responseMap, err := body.GetObject()
```
https://github.com/spyzhov/ajson/blob/0d5fdeed023cf269f55ec781a2713153e3f3ca33/node.go#L494


Alternatively, if you want to deal with byte data directly avoiding ajson library, you could use the same UnmarshalJSON you use above but for map type:
```
responseMap, err := common.UnmarshalJSON[map[string]any](resp)
```

File: providers/attio/metadata.go
Comment (Cobalt0s) [inline providers/attio/metadata.go:19]: There is a recent `common` error that could fit for use case: `ErrMissingExpectedValues`.
Otherwise, export this error, and maybe even not within connector but from `common` package.

File: test/attio/metadata/metadata.go
Comment (Cobalt0s) [inline test/attio/metadata/metadata.go:18]: I was told to use `slog` lib and avoid `log`.
There is `utils.Fail()` that does that underneath.

**Review by Cobalt0s — COMMENTED** [review]
Hi :wave:, some minor comments

File: providers/attio/metadata.go
Comment (Cobalt0s) [reply providers/attio/metadata.go:111]: You are serving Metadata from API and so far we were adding static `schema.json` for those providers that do not have such APIs. Therefore, no action is needed on this one.

.

**PS**: you can check out pipeliner to see how `schema.json` is created and served.
There is OpenAPI yaml or json commited under:
https://github.com/amp-labs/connectors/tree/main/providers/pipeliner/openapi
The `FileManager` loads it (file.go) and is referenced by 2 places:
* Schemas are objects object representation of that file and is used by `ListObjectMetadata` https://github.com/amp-labs/connectors/blob/main/providers/pipeliner/metadata.go.
* And most importantly this is the script which will differ among providers: https://github.com/amp-labs/connectors/blob/main/scripts/openapi/pipeliner/metadata/main.go. It loads OpenAPI file, processes it in custom way and outputs the `schema.json`.

In rare cases `schema.json` is created manually or a website is scrapped, which should be avoided for maintenance cost: https://github.com/amp-labs/connectors/tree/main/scripts/scrapper/salesloft/metadata

Here are some READMEs that I wrote:
* https://github.com/amp-labs/connectors/tree/main/scripts/openapi
* https://github.com/amp-labs/connectors/tree/main/tools/fileconv/api3

**Comment by sanjay-a01** [comment]
updated unmarshal logic, used slog for error handling and common package vars

**Review by Cobalt0s — APPROVED** [review]
Everything is great! Some minor follow up comments.

Additionally, could you add mock tests in this or other PR.
For that, please record sample outputs of `c.Client.Get(ctx, url.String())` from Attio including successful response and some kind of failure. Mock tests would provide freedom to make changes to any code knowing that it will have our back and ensure that it functions as you intended.

**Comment by sanjay-a01** [comment]
- Removed code related to the self object ( Not supported as it is a single object and related to Auth info)
- Added mock test for a list of getting metadata.
- Implemented read functionality in the Attio connector

File: providers/attio/metadata_test.go
Comment (Cobalt0s) [inline providers/attio/metadata_test.go:34]: This is a great test!

File: providers/attio/objectNames.go
Comment (Cobalt0s) [inline providers/attio/objectNames.go:16]: comment is redundant

File: providers/attio/parse.go
Comment (Cobalt0s) [inline providers/attio/parse.go:17]: This method is similar to `common.GetRecordsUnderJSONPath()`. The only difference is that your version is ok with array being null as indicated by the `true` flag.

Maybe there should be 2 methods under common:
* GetRecordsUnderJSONPath -> which would be your version
* MustRecordsUnderJSONPath -> which is current impl GetRecordsUnderJSONPath.

Frankly, I am not sure that distinction is important!
I think before when I was refactoring, if key was not found that indicated that our code is at fault having wrong expectations in payload, thats about it... I have a feeling that you could change `GetRecordsUnderJSONPath` to have a `true` flag and use that method.

File: providers/attio/read.go
Comment (Cobalt0s) [inline providers/attio/read.go:37]: As suggested by above, something along the lines of `common.GetRecordsUnderJSONPath("data")`

**Review by Cobalt0s — APPROVED** [review]
Looks awesome! Thanks for adding so many mock responses, this is good.

@RajatPawar please have a look

**Comment by sanjay-a01** [comment]
Used common functionality for unmarshalling the read response data

File: providers/attio/write.go
Comment (Cobalt0s) [inline providers/attio/write.go:48]: I think we can return success but with data as nil.

@RajatPawar what do you think about this suggestion? Also, when I was going over deep connectors I started wondering can common.WriteResult just have Success field?

**Review by Cobalt0s — APPROVED** [review]
Write functionality looks good to me!

The PR title should be renamed.

Also it would be nice to have multiple dedicate PRs. Write and ListObjectMetadata are fully independent from Read.

File: providers/attio/connector.go
Comment (RajatPawar) [inline providers/attio/connector.go:11]: Can you truncate the `/` out and use a method like filepath.Join (or if we have some other util function) instead wherever we use it? This is easy to make changes and cause bugs with

File: providers/attio/connector.go
Comment (RajatPawar) [inline providers/attio/connector.go:31]: Can you use `Provider()` wherever we want Attio? Not a big problem right now, but it'll help in realizing common code paths and generalizing them

File: providers/attio/parse.go
Comment (RajatPawar) [inline providers/attio/parse.go:8]: Can you move this over to `read`? Would be good to have lesser files whenever possible

File: providers/attio/types.go
Comment (RajatPawar) [inline providers/attio/types.go:8]: I'd move all these to `write.go` for the same rationale as my other comment, especially since they also are all about writing

File: providers/attio/write.go
Comment (RajatPawar) [reply providers/attio/write.go:48]: Looking at other connectors, like @Cobalt0s said, this should probably be an empty but successful `WriteResult`

File: providers/attio/write.go
Comment (RajatPawar) [inline providers/attio/write.go:10]:
```suggestion
// Write creates/updates records in attio. Write currently supports operations to the leads API only.
```

File: providers/attio/metadata.go
Comment (RajatPawar) [inline providers/attio/metadata.go:31]: If there's a way to limit the response to just 1 object, that'd be great. We can avoid waiting on a larger response size

File: providers/attio/README.md
Comment (RajatPawar) [inline providers/attio/README.md:9]: Write seems to be missing as of now?

File: providers/attio/README.md
Comment (RajatPawar) [inline providers/attio/README.md:5]:
```suggestion
Below is an exhaustive list of objects & methods supported on the objects -
```

**Review by RajatPawar — APPROVED** [review]
Approving this, please - 
1. Reply/resolve all the comments before merging it.
2. Attach screenshots of all the tests you've added under a 'Testing' section in the PR description
3. Wherever the test does not actually make calls to the API, please attach a screenshot of a live read/write/metadata call.

Other than that, the PR looks great! Congrats on getting your first connector in 🎉

**Comment by sanjay-a01** [comment]
### Getting a List of Metadata for the supporting objects
image

### Test Result for list of Metadata
image

### Read Functionality Responses
image
image
image
image

### Write Functionality Responses
image
image
image
image

### About creating Webhooks
"Target Url" is one of the body parameters used to create the webhook.
To make the target URL, I created a gin server and implemented the POST method to handle the URL.  Then using ngrok to expose my local server and get the public URL.
With the use of the URL, we can create the webhook. While the Event is triggered, the webhook URL automatically triggered by Attio server.
Also, we can monitor the changes in the ngrok dashboard page.

### Adding Query parameter in read functionality
Some supporting objects have pagination in the Attio connector but, we give the pagination value through a query parameter. So I added the url.WithQueryParams() function manually in the read functionality. If the object has a process to handle the query parameter then the query value will be added to the URL or otherwise, it will neglected.

File: providers/attio/metadata.go
Comment (Jkarage) [inline providers/attio/metadata.go:1]: Hey @sanjay-a01 ,Is there a reason for disabling linter in these files: metadata, ObjectNames, read, write?

File: providers/attio/metadata.go
Comment (Jkarage) [inline providers/attio/metadata.go:12]: I would have named this Data instead of Result.

File: providers/attio/metadata.go
Comment (sanjay-a01) [reply providers/attio/metadata.go:1]: Yeah, I faced the lint issue while raising the PR so only I have disabled the linter.

File: providers/attio/metadata.go
Comment (sanjay-a01) [reply providers/attio/metadata.go:12]: If we specify the Result instead of Data, when we are unmarshalling the response for getting metadata it can't unmarshal because the response data contains "data" as the key.
image

File: providers/attio/write.go
Comment (Jkarage) [inline providers/attio/write.go:39]:
```suggestion
		// uses patch in updating 'single' resource.
```

File: providers/attio/write.go
Comment (Jkarage) [inline providers/attio/write.go:18]:
```suggestion
// Write creates/updates records in attio.
```

File: providers/attio/metadata.go
Comment (Cobalt0s) [reply providers/attio/metadata.go:1]: Yeah, I dont think it should be disabled entirely. Some things are to be addressed and some should be ignored line by line. Otherwise, if more inconsistencies are added the linter would not report them, which is not good.

File: providers/attio/connector.go
Comment (Cobalt0s) [reply providers/attio/connector.go]: `urlbuilder` deals with this issues. It is prefered to not have slashes for sure for readability.
```
// Constructor handles path
urlbuilder.New(c.BaseURL, uriPart1, uriPart2, uriPart3, uriPart4)
```

File: providers/attio/parse.go
Comment (Cobalt0s) [reply providers/attio/parse.go]: This function should belong in common. I can see other connectors using it. As a default behaviour -- no pagination.

File: providers/attio/metadata.go
Comment (Cobalt0s) [reply providers/attio/metadata.go:30]: I think I saw a code somewhere that you wrote that does gorotuines, generalizing it and amplying here would be nice.
Requesting in multithreaded fashion.

File: providers/attio/README.md
Comment (Cobalt0s) [reply providers/attio/README.md:9]: I think objectNames.go is an exact representation of this README. I would remove README, maybe would add comments to that file instead. There is always an incentive to keep go code up to date unlike `.md`, `.txt` type of files.

File: providers/attio/metadata.go
Comment (Cobalt0s) [reply providers/attio/metadata.go]: I think the suggestion is 
```
Data []map[string]any `json:"data"`
```
if understood correctly.

File: providers/attio/read.go
Comment (Cobalt0s) [inline providers/attio/read.go:29]: This doesn't seem right. This is correct during development process. We don't want to query only 1 item at a time!
Also if some objects have offset pagination support that should be implemented instead of `dummyNextPageFunc`.

**Review by Cobalt0s — CHANGES_REQUESTED** [review]
There is an action needed regarding `read.go`. You need to handle pagination or at least remove `per page limit of 1`.

**Comment by sanjay-a01** [comment]
I have changed the pagination logic for the read functionality in the Attio connector

image

In that while using **DumpJSON**  function for printing the response, I faced issue with the next page response like 

image

and also I tried with **Json.MarshalIndent** function again getting the same result.
so I changed the marshalIndent logic like below

image
image

and got the nextpage URL like

image

It is just printing the response so I removed my marshalIndent logic and used **json.MarshalIndent** function for printing the response.

some object in the Attio connector don't require pagination, so I make the restriction using supportLimitAndOffset variable and check in the read functionality like

image
image

If the URL contains query parameters like limit and offset and has a response value, then only next page data will generate in the response otherwise it will return empty value in the next page

image

File: providers/attio/parse.go
Comment (Cobalt0s) [inline providers/attio/parse.go:23]: I think it is worth changing urlbuilder.URL to address this issue instead of converting to url.URL.
urlbuilder holds query parameters already:
```
type URL struct {
	delegate           *url.URL
	queryParams        url.Values
	encodingExceptions map[string]string
}
```
Just add `HasQueryParam(name string)` method and delegate the call `u.queryParams.Has(name)`.

File: providers/attio/parse.go
Comment (Cobalt0s) [inline providers/attio/parse.go:39]:
```suggestion
		}
		
		return "", nil
```
Else keyword is not needed due to `return` under if statement.

**Review by Cobalt0s — CHANGES_REQUESTED** [review]
Everything is great! I see lots of effort and well done!

One important thing is check the comments for `makeNextRecordsURL` method.
You should update `urlbuilder` package. And make use of `jsonquery` to avoid `panic()` calls.

File: providers/attio/params.go
Comment (sanjay-a01) [reply providers/attio/params.go:21]: yes, I have added the default size which is mentioned in the attio documentation

File: providers/attio/parse.go
Comment (sanjay-a01) [reply providers/attio/parse.go]: I removed dummyNextPageFunc function because some object in Attio connector has pagination. Instead of this function, I created new function makeNextRecordsURL function in attio/parse.go file.

**Review by Cobalt0s — APPROVED** [review]
Awesome!
LGTM

**Pull request merged**
